pyfr/backends/openmp/types.py

```diff
@@ -77,13 +87,62 @@ def commit(self):
         self._runlist, i = [], 0
 
         for j in sorted(self.mpi_idxs):
-            krunargs = make_array(k.runargs for k in self.klist[i:j])
-            self._runlist.append((j - i, krunargs, self.mpi_idxs[j]))
+            self._runlist_update(i, j, self.mpi_idxs[j])
             i = j
 
         if i != len(self.klist) - 1:
-            krunargs = make_array(k.runargs for k in self.klist[i:])
-            self._runlist.append((len(self.klist) - i, krunargs, []))
+            self._runlist_update(i, len(self.klist), [])
+
+    def group(self, knodes, subs=[]):
+        self.kgnodes.append([k.kernel for kernel in knodes for k in kernel])
+
+        # Loop over kernel groups per each element type in the domain
+        for ks in zip(*knodes):
+            self.bkargs.append([])
+            bka = []
+            for k in ks:
+                kernel = k.kernel
+                fun = cast(kernel.fun, c_void_p)
+                args, argsz = addressof(kernel.kargs), sizeof(kernel.kargs)
+
+                bka.append({'fun':fun, 'args':args, 'argsz':argsz})
+
+            nsubs = 0
+            argsubs = []
+            allocsz = 0
+
+            # Loop over entries in substitutions
+            for elmgr in subs:
+                nsubs += len(elmgr)
+                for ka_pair in elmgr:
+                    krnl, arg = ka_pair
+                    for i, k in enumerate(ks):
+                        if (krnl == k.kernel.fname):
+                            offset = k.kernel.arg_off(arg)
+                            ix = k.kernel.argn.index(arg)
+                            blksz = k.kernel.argblks[f'arg{ix}']
+                            argsubs.append([i, offset, allocsz])
+
+                            bka[i]['argmask'] = ix
+
+                allocsz += blksz
+
+            c_argsubs = make_array(
+                np.array(argsubs, dtype=np.int32).flatten(), type=c_int
+            )
+
+            self.bkargs.append([OpenMPBlockKernelArgs(**d) for d in bka])
+            self.kptrs = make_array(self.bkargs[-1])
+
+            self.b_runargs = OpenMPBlockRunArgs(
```

Comment:
I don't think you need to store this as it will be copied over when you construct the OpenMPKRunArgs (as it is by value not by pointer).